staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go

```diff
@@ -31,171 +31,278 @@ import (
+	}
+	return nil, false
 }
 
-func getNestedMap(obj map[string]interface{}, fields ...string) map[string]string {
-	if m, ok := getNestedField(obj, fields...).(map[string]interface{}); ok {
+// NestedStringMap returns the map[string]string value of a nested field.
+// Returns false if value is not found or is not a map[string]interface{}.
+func NestedStringMap(obj map[string]interface{}, fields ...string) (map[string]string, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.(map[string]interface{}); ok {
 		strMap := make(map[string]string, len(m))
 		for k, v := range m {
 			if str, ok := v.(string); ok {
 				strMap[k] = str
+			} else {
+				return nil, false
 			}
 		}
-		return strMap
+		return strMap, true
+	}
+	return nil, false
+}
+
+// NestedMap returns the map[string]interface{} value of a nested field.
+// Returns false if value is not found or is not a map[string]interface{}.
+func NestedMap(obj map[string]interface{}, fields ...string) (map[string]interface{}, bool) {
+	val, ok := NestedField(obj, fields...)
+	if !ok {
+		return nil, false
+	}
+	if m, ok := val.(map[string]interface{}); ok {
+		nMap := make(map[string]interface{}, len(m))
+		for k, v := range m {
+			nMap[k] = v
+		}
+		return nMap, true
+	}
+	return nil, false
+}
+
+func assertValueType(val interface{}) {
+	switch x := val.(type) {
+	case map[string]interface{}:
+		for _, v := range x {
+			assertValueType(v)
+		}
+	case []interface{}:
+		for _, v := range x {
+			assertValueType(v)
+		}
+	case string, int64, bool, float64, nil, gojson.Number:
+	default:
+		panic(fmt.Errorf("invalid value type: %T", val))
 	}
-	return nil
 }
 
-func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+// SetNestedField sets the value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedField(obj map[string]interface{}, value interface{}, fields ...string) bool {
+	assertValueType(value)
 	m := obj
-	if len(fields) > 1 {
-		for _, field := range fields[0 : len(fields)-1] {
-			if _, ok := m[field].(map[string]interface{}); !ok {
-				m[field] = make(map[string]interface{})
+	for _, field := range fields[:len(fields)-1] {
+		if val, ok := m[field]; ok {
+			if valMap, ok := val.(map[string]interface{}); ok {
+				m = valMap
+			} else {
+				return false
 			}
-			m = m[field].(map[string]interface{})
+		} else {
+			newVal := make(map[string]interface{})
+			m[field] = newVal
+			m = newVal
 		}
 	}
 	m[fields[len(fields)-1]] = value
+	return true
 }
 
-func setNestedSlice(obj map[string]interface{}, value []string, fields ...string) {
+// SetNestedStringSlice sets the string slice value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedStringSlice(obj map[string]interface{}, value []string, fields ...string) bool {
```

Comment:
For the setters the bool return value feels unnatural.

Reply:
Does an explicit error here look better? Something like that in https://github.com/kubernetes/kubernetes/pull/55168#discussion_r152580964?